---
title: "Comparing common and rare single variant and gene aggregate instrumentation strategies for molecular MR"
author: "Aimee Hanson"
date: "`r format(Sys.time(), '%d %B %Y')`"
format: 
  html:
    toc: true
    toc-depth: 4
    toc-location: left
    page-layout: article
    code-fold: true
---

```{r}
#| label: setup
#| message: FALSE
#| echo: FALSE

library(here)
library(dplyr)
library(ggplot2)
library(dotenv)

dotenv::load_dot_env(file = here("config.env"))
data_dir <- Sys.getenv("data_dir")
res_dir <- Sys.getenv("results_dir")

knitr::opts_chunk$set(message = FALSE, warning = FALSE)

source(here("scripts/04_analysis_proteins/functions.R"))
```
## Introduction

...


## Data import

Read in MR results from `scripts/04_analysis_proteins/001_performmolecularMR.R` and harmonised summary statistics:

```{r}
#| label: import mr results

files <- list.files(res_dir, pattern = "results_molecular", full.names = T)

# Import MR results
res_MR <- list()
for(i in 1:length(files)){
  load(files[i])
  res_MR[[i]] <- results_MR
}
names(res_MR) <- sub(".*_MR_(.*)\\.rda","\\1", files)


load(file = file.path(res_dir, "results_complextrait_MR.rda"))
# Import harmonised study data
load(file = file.path(data_dir, "harmonised", "harmonised_studies.rda"))

# Retain exposure-outcome pairs with significant IVW estimate for at least one instrumentation strategy
res_MR <- lapply(results_MR, function(x){
  sig_ivw <- x |> filter(method == "Inverse variance weighted" & pval < 0.05)
  if(nrow(sig_ivw) == 0){
    return(NULL)
  } else {
    return(x)
  }
}) 

# 2937 proteins tested per outcome
alpha <- signif(0.05/length(res_MR$BMI),2)
```

## Summary of findings

The number of proteins with significant causal estimates across outcomes and instrumentation strategies

```{r}
#| label: summary
#| warning: FALSE
#| fig.width: 12
#| fig.height: 3

variant_IVsets <- c("dhindsa_exwas_variant_common" = "exome:common",
                    "dhindsa_exwas_variant_rare" = "exome:rare",
                    "dhindsa_exwas_variant_ultrarare" = "exome:ultrarare",
                    "sun_gwas_variant_common_clumped" = "genome:common (clumped)",
                    "sun_gwas_variant_common_finemapped" = "genome:common (fmapped)")

cols_variant_IVsets <- c("steelblue4","steelblue3","steelblue1","darkorange","orange")
names(cols_variant_IVsets) <- variant_IVsets

# Number of proteins with significant causal estimates across instrument sets
summary_sig <- lapply(res_MR, summarise_results, p_thresh = alpha, n_snps = 3)

# Number of proteins with instrument sets (of >= n_snps) available (shown for BMI - differs slightly across outcomes depending on whether IVs are available in outcome study)
avail_IVs <- do.call("rbind",lapply(summary_sig, function(x){apply(x$min_p, MARGIN = 2, function(y){sum(!is.na(y))})/nrow(x$min_p) * 100}))

avail_IVs[,names(variant_IVsets)] |> 
  as.data.frame() |> 
  tibble::rownames_to_column("outcome") |>
  tidyr::pivot_longer(cols = 2:6) |>
  dplyr::mutate(label = variant_IVsets[name]) |>
  ggplot(aes(x = label, y = value, fill = label)) +
  facet_grid(.~outcome) +
  geom_col() +
  scale_fill_manual(values = cols_variant_IVsets) +
  theme_minimal() +
  theme(axis.text.x = element_blank(), 
        axis.ticks.x = element_blank(),
        plot.title = element_text(hjust = 0.5),
        legend.position = "bottom") +
  labs(y = "%", title = "Percentage of O-link proteins with IV sets available (with n>=3 instruments)")

### Repeat for cis and trans instruments


# Proportion of causally associated protein across those with instruments available

check <- summary_sig$BMI$thresh_p



res_MR_top <- lapply(res_MR, filter_results, p_thresh = alpha, n_snps = 3)











```


Retaining exposure --> outcome pairs where all variant-based IV sets have >=4 genetic instruments and at least one gives a significant causal estimate at p < `r alpha`.




