<!DOCTYPE html>
<html xmlns="http://www.w3.org/1999/xhtml" lang="en" xml:lang="en"><head>

<meta charset="utf-8">
<meta name="generator" content="quarto-1.6.40">

<meta name="viewport" content="width=device-width, initial-scale=1.0, user-scalable=yes">

<meta name="author" content="Aimee Hanson">
<meta name="dcterms.date" content="2025-09-16">

<title>Comparing common and rare single variant and gene aggregate instrumentation strategies for molecular MR</title>
<style>
code{white-space: pre-wrap;}
span.smallcaps{font-variant: small-caps;}
div.columns{display: flex; gap: min(4vw, 1.5em);}
div.column{flex: auto; overflow-x: auto;}
div.hanging-indent{margin-left: 1.5em; text-indent: -1.5em;}
ul.task-list{list-style: none;}
ul.task-list li input[type="checkbox"] {
  width: 0.8em;
  margin: 0 0.8em 0.2em -1em; /* quarto-specific, see https://github.com/quarto-dev/quarto-cli/issues/4556 */ 
  vertical-align: middle;
}
/* CSS for syntax highlighting */
pre > code.sourceCode { white-space: pre; position: relative; }
pre > code.sourceCode > span { line-height: 1.25; }
pre > code.sourceCode > span:empty { height: 1.2em; }
.sourceCode { overflow: visible; }
code.sourceCode > span { color: inherit; text-decoration: inherit; }
div.sourceCode { margin: 1em 0; }
pre.sourceCode { margin: 0; }
@media screen {
div.sourceCode { overflow: auto; }
}
@media print {
pre > code.sourceCode { white-space: pre-wrap; }
pre > code.sourceCode > span { display: inline-block; text-indent: -5em; padding-left: 5em; }
}
pre.numberSource code
  { counter-reset: source-line 0; }
pre.numberSource code > span
  { position: relative; left: -4em; counter-increment: source-line; }
pre.numberSource code > span > a:first-child::before
  { content: counter(source-line);
    position: relative; left: -1em; text-align: right; vertical-align: baseline;
    border: none; display: inline-block;
    -webkit-touch-callout: none; -webkit-user-select: none;
    -khtml-user-select: none; -moz-user-select: none;
    -ms-user-select: none; user-select: none;
    padding: 0 4px; width: 4em;
  }
pre.numberSource { margin-left: 3em;  padding-left: 4px; }
div.sourceCode
  {   }
@media screen {
pre > code.sourceCode > span > a:first-child::before { text-decoration: underline; }
}
</style>


<script src="rarevcommon_molecularMR_files/libs/clipboard/clipboard.min.js"></script>
<script src="rarevcommon_molecularMR_files/libs/quarto-html/quarto.js"></script>
<script src="rarevcommon_molecularMR_files/libs/quarto-html/popper.min.js"></script>
<script src="rarevcommon_molecularMR_files/libs/quarto-html/tippy.umd.min.js"></script>
<script src="rarevcommon_molecularMR_files/libs/quarto-html/anchor.min.js"></script>
<link href="rarevcommon_molecularMR_files/libs/quarto-html/tippy.css" rel="stylesheet">
<link href="rarevcommon_molecularMR_files/libs/quarto-html/quarto-syntax-highlighting-549806ee2085284f45b00abea8c6df48.css" rel="stylesheet" id="quarto-text-highlighting-styles">
<script src="rarevcommon_molecularMR_files/libs/bootstrap/bootstrap.min.js"></script>
<link href="rarevcommon_molecularMR_files/libs/bootstrap/bootstrap-icons.css" rel="stylesheet">
<link href="rarevcommon_molecularMR_files/libs/bootstrap/bootstrap-973236bd072d72a04ee9cd82dcc9cb29.min.css" rel="stylesheet" append-hash="true" id="quarto-bootstrap" data-mode="light">


</head>

<body>

<div id="quarto-content" class="page-columns page-rows-contents page-layout-full toc-left">
<div id="quarto-sidebar-toc-left" class="sidebar toc-left">
  <nav id="TOC" role="doc-toc" class="toc-active">
    <h2 id="toc-title">Table of contents</h2>
   
  <ul>
  <li><a href="#introduction" id="toc-introduction" class="nav-link active" data-scroll-target="#introduction">Introduction</a>
  <ul>
  <li><a href="#instrumental-variable-iv-sets" id="toc-instrumental-variable-iv-sets" class="nav-link" data-scroll-target="#instrumental-variable-iv-sets">Instrumental variable (IV) sets</a></li>
  </ul></li>
  <li><a href="#data-import" id="toc-data-import" class="nav-link" data-scroll-target="#data-import">Data import</a></li>
  <li><a href="#comparison-of-iv-sets" id="toc-comparison-of-iv-sets" class="nav-link" data-scroll-target="#comparison-of-iv-sets">Comparison of IV sets</a></li>
  <li><a href="#next-steps" id="toc-next-steps" class="nav-link" data-scroll-target="#next-steps">Next steps</a></li>
  </ul>
</nav>
</div>
<div id="quarto-margin-sidebar" class="sidebar margin-sidebar zindex-bottom">
</div>
<main class="content column-page-right" id="quarto-document-content">

<header id="title-block-header" class="quarto-title-block default">
<div class="quarto-title">
<h1 class="title">Comparing common and rare single variant and gene aggregate instrumentation strategies for molecular MR</h1>
</div>



<div class="quarto-title-meta column-page-right">

    <div>
    <div class="quarto-title-meta-heading">Author</div>
    <div class="quarto-title-meta-contents">
             <p>Aimee Hanson </p>
          </div>
  </div>
    
    <div>
    <div class="quarto-title-meta-heading">Published</div>
    <div class="quarto-title-meta-contents">
      <p class="date">September 16, 2025</p>
    </div>
  </div>
  
    
  </div>
  


</header>


<section id="introduction" class="level2">
<h2 class="anchored" data-anchor-id="introduction">Introduction</h2>
<p>Classically, Mendelian Randomisation (MR) methods utilising trait-associated genetic variants from GWAS studies have employed common polymorphisms (population MAF &gt; 1%) targeted by genotyping chips, or reliably imputed from reference populations, to instrument modifiable exposures. However, common variants tested in GWAS typically explain a very small fraction of the variability in a measured trait, potentially exhibit pleiotropic effects acted upon by balancing selection or as a consequent of genetic linkage, and are rarely causal. Rare variants, which typically show large biological effects (e.g.&nbsp;through abolishing protein expression) provide a means of more unambiguously instrumenting relevant molecular processes. Comparison of causal estimates derived using differing methods of genetically instrumenting modifiable exposures may enhance the interpretation of the biological mechanisms underlying exposure-outcome relationships. This includes using variants from across the allele frequency spectrum, but also leveraging rare variant aggregate approaches to instrument gene-level perturbations in expression and function.</p>
<p>Here, caual estimates for pairwises combinations of <strong>2937</strong> molecular exposures (O-link plasma proteins) and <strong>11</strong> health related compelex traits (below) have been derived using nine instrumentation strategies, with further instrument splitting into cis (variants within 1MB of the encoding gene start and end site), trans and full variant sets (a total of <strong>27</strong> instrument sets per exposure).</p>
<p><strong>Outcomes:</strong> Low density lipoprotein levels (LDL), Body Mass Index (BMI), Vitamin D (VitD), Triglycerides (Trig), Glycated Haemoglobin (HbA1c), Mean Platelet Volume (PlatVol), IGF-1 (IGF1), Waist-to-Hip Ratio (WHRadjBMI), Red Blood Cell Erythrocyte Count (RBCCount), Mean Corpuscular Volume (CorpVol) and Systolic Blood Pressure (SBP)</p>
<section id="instrumental-variable-iv-sets" class="level3">
<h3 class="anchored" data-anchor-id="instrumental-variable-iv-sets">Instrumental variable (IV) sets</h3>
<p>Nine sets of IVs for each protein exposure have been extracted from across two studies:<br>
GWAS common variants: “Plasma proteomic associations with genetics and health in the UK Biobank” (Sun, B. 2023, Nature).<br>
WES variants/gene aggregate tests: “Rare variant associations with plasma protein levels in the UK Biobank” (Dhindsa, R. 2023, Nature).</p>
<p><strong>Common GWAS</strong><br>
Common genome-wide fine-mapped variants from the GPMAP (&gt;1% MAF) (cis, trans and combined)<br>
Common genome-wide clumped and pruned variants (&gt;1% MAF) (cis, trans and combined)</p>
<p><strong>WES (variants)</strong><br>
Common exome-wide (&gt;1% MAF) (cis, trans and combined)<br>
Rare exome-wide (0-1% MAF) (cis, trans and combined)<br>
Ultra-rare exome-wide (0-0.1% MAF) (cis, trans and combined)<br>
<em>LD clumping was performed for common variants (for which LD matrices have been derived) only. For rare variants, instrument sets were stored both pre and post filtering to the top associated variant per gene for sensitivity analyses.</em></p>
<p><strong>WES (masks)</strong><br>
pLOF burden mask (cis, trans and combined)<br>
missense/low-confidence burden mask (cis, trans and combined)<br>
pLOF/missense/low-confidence burden mask (cis, trans and combined)<br>
synonymous burden mask (cis, trans and combined)</p>
</section>
</section>
<section id="data-import" class="level2">
<h2 class="anchored" data-anchor-id="data-import">Data import</h2>
<p>Read in MR results from <code>scripts/04_analysis_proteins/001_performmolecularMR.R</code> and harmonised summary statistics:</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb1"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb1-1"><a href="#cb1-1" aria-hidden="true" tabindex="-1"></a>files <span class="ot">&lt;-</span> <span class="fu">list.files</span>(res_dir, <span class="at">pattern =</span> <span class="st">"results_molecular"</span>, <span class="at">full.names =</span> T)</span>
<span id="cb1-2"><a href="#cb1-2" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-3"><a href="#cb1-3" aria-hidden="true" tabindex="-1"></a><span class="co"># Import MR results</span></span>
<span id="cb1-4"><a href="#cb1-4" aria-hidden="true" tabindex="-1"></a>res_MR <span class="ot">&lt;-</span> <span class="fu">list</span>()</span>
<span id="cb1-5"><a href="#cb1-5" aria-hidden="true" tabindex="-1"></a><span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(files)){</span>
<span id="cb1-6"><a href="#cb1-6" aria-hidden="true" tabindex="-1"></a>  <span class="fu">load</span>(files[i])</span>
<span id="cb1-7"><a href="#cb1-7" aria-hidden="true" tabindex="-1"></a>  res_MR[[i]] <span class="ot">&lt;-</span> results_MR</span>
<span id="cb1-8"><a href="#cb1-8" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb1-9"><a href="#cb1-9" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(res_MR) <span class="ot">&lt;-</span> <span class="fu">sub</span>(<span class="st">".*_MR_(.*)</span><span class="sc">\\</span><span class="st">.rda"</span>,<span class="st">"</span><span class="sc">\\</span><span class="st">1"</span>, files)</span>
<span id="cb1-10"><a href="#cb1-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-11"><a href="#cb1-11" aria-hidden="true" tabindex="-1"></a><span class="co"># Retain exposure-outcome pairs with significant IVW estimate for at least one instrumentation strategy</span></span>
<span id="cb1-12"><a href="#cb1-12" aria-hidden="true" tabindex="-1"></a><span class="co"># res_MR_top &lt;- lapply(res_MR, function(x){</span></span>
<span id="cb1-13"><a href="#cb1-13" aria-hidden="true" tabindex="-1"></a><span class="co">#   sig_ivw &lt;- x |&gt; filter(method == "Inverse variance weighted" &amp; pval &lt; 0.05)</span></span>
<span id="cb1-14"><a href="#cb1-14" aria-hidden="true" tabindex="-1"></a><span class="co">#   if(nrow(sig_ivw) == 0){</span></span>
<span id="cb1-15"><a href="#cb1-15" aria-hidden="true" tabindex="-1"></a><span class="co">#     return(NULL)</span></span>
<span id="cb1-16"><a href="#cb1-16" aria-hidden="true" tabindex="-1"></a><span class="co">#   } else {</span></span>
<span id="cb1-17"><a href="#cb1-17" aria-hidden="true" tabindex="-1"></a><span class="co">#     return(x)</span></span>
<span id="cb1-18"><a href="#cb1-18" aria-hidden="true" tabindex="-1"></a><span class="co">#   }</span></span>
<span id="cb1-19"><a href="#cb1-19" aria-hidden="true" tabindex="-1"></a><span class="co"># }) </span></span>
<span id="cb1-20"><a href="#cb1-20" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-21"><a href="#cb1-21" aria-hidden="true" tabindex="-1"></a><span class="co"># 0.05/2937 proteins tested per outcome</span></span>
<span id="cb1-22"><a href="#cb1-22" aria-hidden="true" tabindex="-1"></a>alpha <span class="ot">&lt;-</span> <span class="fu">signif</span>(<span class="fl">0.05</span><span class="sc">/</span><span class="fu">length</span>(res_MR<span class="sc">$</span>BMI),<span class="dv">2</span>)</span>
<span id="cb1-23"><a href="#cb1-23" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-24"><a href="#cb1-24" aria-hidden="true" tabindex="-1"></a><span class="co"># IVsets</span></span>
<span id="cb1-25"><a href="#cb1-25" aria-hidden="true" tabindex="-1"></a>IVsets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"dhindsa_exwas_mask_ptv"</span> <span class="ot">=</span> <span class="st">"mask:pLOF"</span>,</span>
<span id="cb1-26"><a href="#cb1-26" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_mask_raredmg"</span> <span class="ot">=</span> <span class="st">"mask:missense"</span>,</span>
<span id="cb1-27"><a href="#cb1-27" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_mask_ptvraredmg"</span> <span class="ot">=</span> <span class="st">"mask:pLOF|missense"</span>,</span>
<span id="cb1-28"><a href="#cb1-28" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_mask_syn"</span> <span class="ot">=</span> <span class="st">"mask:synonymous"</span>,</span>
<span id="cb1-29"><a href="#cb1-29" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_variant_rare"</span> <span class="ot">=</span> <span class="st">"exome:rare"</span>,</span>
<span id="cb1-30"><a href="#cb1-30" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_variant_rare_filt"</span> <span class="ot">=</span> <span class="st">"exome:rare (filtered)"</span>,</span>
<span id="cb1-31"><a href="#cb1-31" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_variant_ultrarare"</span> <span class="ot">=</span> <span class="st">"exome:ultrarare"</span>,</span>
<span id="cb1-32"><a href="#cb1-32" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_variant_ultrarare_filt"</span> <span class="ot">=</span> <span class="st">"exome:ultrarare (filtered)"</span>,</span>
<span id="cb1-33"><a href="#cb1-33" aria-hidden="true" tabindex="-1"></a>            <span class="st">"dhindsa_exwas_variant_common"</span> <span class="ot">=</span> <span class="st">"exome:common (clumped)"</span>,</span>
<span id="cb1-34"><a href="#cb1-34" aria-hidden="true" tabindex="-1"></a>            <span class="st">"sun_gwas_variant_common_clumped"</span> <span class="ot">=</span> <span class="st">"genome:common (clumped)"</span>,</span>
<span id="cb1-35"><a href="#cb1-35" aria-hidden="true" tabindex="-1"></a>            <span class="st">"sun_gwas_variant_common_finemapped"</span> <span class="ot">=</span> <span class="st">"genome:common (fmapped)"</span>)</span>
<span id="cb1-36"><a href="#cb1-36" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-37"><a href="#cb1-37" aria-hidden="true" tabindex="-1"></a>cols_IVsets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"darkorchid4"</span>,<span class="st">"darkorchid3"</span>,<span class="st">"darkorchid2"</span>,<span class="st">"orchid"</span>,</span>
<span id="cb1-38"><a href="#cb1-38" aria-hidden="true" tabindex="-1"></a>                 <span class="st">"steelblue4"</span>,<span class="st">"steelblue4"</span>,<span class="st">"steelblue1"</span>,<span class="st">"steelblue1"</span>,<span class="st">"darkorange2"</span>,<span class="st">"orange"</span>,<span class="st">"orange"</span>)</span>
<span id="cb1-39"><a href="#cb1-39" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cols_IVsets) <span class="ot">&lt;-</span> IVsets</span>
<span id="cb1-40"><a href="#cb1-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-41"><a href="#cb1-41" aria-hidden="true" tabindex="-1"></a><span class="co"># Variant-based instrument sets</span></span>
<span id="cb1-42"><a href="#cb1-42" aria-hidden="true" tabindex="-1"></a>variant_IVsets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"dhindsa_exwas_variant_rare"</span> <span class="ot">=</span> <span class="st">"exome:rare"</span>,</span>
<span id="cb1-43"><a href="#cb1-43" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_rare_filt"</span> <span class="ot">=</span> <span class="st">"exome:rare (filtered)"</span>,</span>
<span id="cb1-44"><a href="#cb1-44" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_ultrarare"</span> <span class="ot">=</span> <span class="st">"exome:ultrarare"</span>,</span>
<span id="cb1-45"><a href="#cb1-45" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_ultrarare_filt"</span> <span class="ot">=</span> <span class="st">"exome:ultrarare (filtered)"</span>,</span>
<span id="cb1-46"><a href="#cb1-46" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_common"</span> <span class="ot">=</span> <span class="st">"exome:common (clumped)"</span>,</span>
<span id="cb1-47"><a href="#cb1-47" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"sun_gwas_variant_common_clumped"</span> <span class="ot">=</span> <span class="st">"genome:common (clumped)"</span>,</span>
<span id="cb1-48"><a href="#cb1-48" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"sun_gwas_variant_common_finemapped"</span> <span class="ot">=</span> <span class="st">"genome:common (fmapped)"</span>)</span>
<span id="cb1-49"><a href="#cb1-49" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-50"><a href="#cb1-50" aria-hidden="true" tabindex="-1"></a>cols_variant_IVsets <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"steelblue4"</span>,<span class="st">"steelblue4"</span>,<span class="st">"steelblue1"</span>,<span class="st">"steelblue1"</span>,<span class="st">"darkorange2"</span>,<span class="st">"orange"</span>,<span class="st">"orange"</span>)</span>
<span id="cb1-51"><a href="#cb1-51" aria-hidden="true" tabindex="-1"></a><span class="fu">names</span>(cols_variant_IVsets) <span class="ot">&lt;-</span> variant_IVsets</span>
<span id="cb1-52"><a href="#cb1-52" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb1-53"><a href="#cb1-53" aria-hidden="true" tabindex="-1"></a><span class="co"># Filter variant-based instruments </span></span>
<span id="cb1-54"><a href="#cb1-54" aria-hidden="true" tabindex="-1"></a>variant_IVsets_indep <span class="ot">&lt;-</span> <span class="fu">c</span>(<span class="st">"dhindsa_exwas_variant_rare_filt"</span> <span class="ot">=</span> <span class="st">"exome:rare (filtered)"</span>,</span>
<span id="cb1-55"><a href="#cb1-55" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_ultrarare_filt"</span> <span class="ot">=</span> <span class="st">"exome:ultrarare (filtered)"</span>,</span>
<span id="cb1-56"><a href="#cb1-56" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"dhindsa_exwas_variant_common"</span> <span class="ot">=</span> <span class="st">"exome:common (clumped)"</span>,</span>
<span id="cb1-57"><a href="#cb1-57" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"sun_gwas_variant_common_clumped"</span> <span class="ot">=</span> <span class="st">"genome:common (clumped)"</span>,</span>
<span id="cb1-58"><a href="#cb1-58" aria-hidden="true" tabindex="-1"></a>                    <span class="st">"sun_gwas_variant_common_finemapped"</span> <span class="ot">=</span> <span class="st">"genome:common (fmapped)"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
</div>
</section>
<section id="comparison-of-iv-sets" class="level2">
<h2 class="anchored" data-anchor-id="comparison-of-iv-sets">Comparison of IV sets</h2>
<p><strong>The average number of genetic instruments (pQTLs) per protein across instrument sets</strong></p>
<p>Rare and common variant instruments (pQTLs) have been partitioned according to their proximity to the instrumented protein-coding gene into cis (+- 1MB from gene boundaries) and trans (&gt;1MB distant) sets. There is a higher average number of exonic rare and ultra-rare variant instruments per protein derived from WES data than common GWAS derived instruments. Proteins that are instrumented with &gt;100 rare cis variants fall in gene dense genomic regions. Available IVs drop when restricting to one rare variant per annotated gene, but rare IVs have not been explicitly LD pruned so it is possible that in some instances these variants are not independent. More common variant trans-IVs can be derived from the WES over the GWAS datasets (with clumping and pruning), potentially due to the higher coverage of coding regions involved in trans-gene regulation (via modification of pathwaway effects) by exome sequencing data.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb2"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb2-1"><a href="#cb2-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of instruments per protein for each IV set</span></span>
<span id="cb2-2"><a href="#cb2-2" aria-hidden="true" tabindex="-1"></a>num_IVs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="cf">function</span>(x){</span>
<span id="cb2-3"><a href="#cb2-3" aria-hidden="true" tabindex="-1"></a>  df_out <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="fu">expand.grid</span>(<span class="at">IV_set =</span> IV_sets, <span class="at">cis_trans =</span> <span class="fu">c</span>(<span class="st">"cis"</span>,<span class="st">"trans"</span>)))</span>
<span id="cb2-4"><a href="#cb2-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-5"><a href="#cb2-5" aria-hidden="true" tabindex="-1"></a>  <span class="cf">for</span>(i <span class="cf">in</span> <span class="dv">1</span><span class="sc">:</span><span class="fu">length</span>(x)){</span>
<span id="cb2-6"><a href="#cb2-6" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> x[[i]] <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="fu">unique</span>(nsnp), <span class="at">.by =</span> <span class="fu">c</span>(IV_set, cis_trans))</span>
<span id="cb2-7"><a href="#cb2-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">colnames</span>(df)[<span class="dv">3</span>] <span class="ot">&lt;-</span> <span class="fu">names</span>(x[i])</span>
<span id="cb2-8"><a href="#cb2-8" aria-hidden="true" tabindex="-1"></a>    df_out <span class="ot">&lt;-</span> <span class="fu">left_join</span>(df_out, df, <span class="at">by =</span> <span class="fu">c</span>(<span class="st">"IV_set"</span>,<span class="st">"cis_trans"</span>))</span>
<span id="cb2-9"><a href="#cb2-9" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb2-10"><a href="#cb2-10" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-11"><a href="#cb2-11" aria-hidden="true" tabindex="-1"></a>  df_out <span class="ot">&lt;-</span> df_out <span class="sc">|&gt;</span> tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="dv">3</span><span class="sc">:</span><span class="fu">length</span>(df_out)) <span class="sc">|&gt;</span> <span class="fu">na.exclude</span>()</span>
<span id="cb2-12"><a href="#cb2-12" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df_out)</span>
<span id="cb2-13"><a href="#cb2-13" aria-hidden="true" tabindex="-1"></a>})</span>
<span id="cb2-14"><a href="#cb2-14" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-15"><a href="#cb2-15" aria-hidden="true" tabindex="-1"></a>num_IVs <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, num_IVs) <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">outcome =</span> <span class="fu">sub</span>(<span class="st">".*_"</span>,<span class="st">""</span>,name))</span>
<span id="cb2-16"><a href="#cb2-16" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-17"><a href="#cb2-17" aria-hidden="true" tabindex="-1"></a>p1 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(<span class="at">FUN =</span> <span class="cf">function</span>(x,y){</span>
<span id="cb2-18"><a href="#cb2-18" aria-hidden="true" tabindex="-1"></a>  x <span class="ot">&lt;-</span> x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(cis_trans <span class="sc">==</span> y) <span class="sc">|&gt;</span></span>
<span id="cb2-19"><a href="#cb2-19" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">factor</span>(IVsets[IV_set], <span class="at">levels =</span> IVsets)) <span class="sc">|&gt;</span></span>
<span id="cb2-20"><a href="#cb2-20" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(label <span class="sc">%in%</span> <span class="fu">grep</span>(<span class="st">"exome|genome"</span>, IVsets, <span class="at">value =</span> T))</span>
<span id="cb2-21"><a href="#cb2-21" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-22"><a href="#cb2-22" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(x, <span class="fu">aes</span>(<span class="at">x =</span> label, <span class="at">y =</span> value, <span class="at">fill =</span> label)) <span class="sc">+</span></span>
<span id="cb2-23"><a href="#cb2-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(.<span class="sc">~</span>outcome) <span class="sc">+</span></span>
<span id="cb2-24"><a href="#cb2-24" aria-hidden="true" tabindex="-1"></a>    <span class="fu">stat_boxplot</span>(<span class="at">geom =</span><span class="st">'errorbar'</span>, <span class="at">width =</span> <span class="fl">0.2</span>) <span class="sc">+</span> </span>
<span id="cb2-25"><a href="#cb2-25" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_boxplot</span>(<span class="at">width =</span> <span class="fl">0.7</span>) <span class="sc">+</span></span>
<span id="cb2-26"><a href="#cb2-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_log10</span>() <span class="sc">+</span></span>
<span id="cb2-27"><a href="#cb2-27" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">values =</span> cols_IVsets) <span class="sc">+</span></span>
<span id="cb2-28"><a href="#cb2-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb2-29"><a href="#cb2-29" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb2-30"><a href="#cb2-30" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.position =</span> <span class="st">"none"</span>,</span>
<span id="cb2-31"><a href="#cb2-31" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>)) <span class="sc">+</span></span>
<span id="cb2-32"><a href="#cb2-32" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"IV set"</span>,</span>
<span id="cb2-33"><a href="#cb2-33" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"Num. instruments"</span>, </span>
<span id="cb2-34"><a href="#cb2-34" aria-hidden="true" tabindex="-1"></a>         <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">"Number of "</span>,y ,<span class="st">" instruments per plasma protein"</span>))</span>
<span id="cb2-35"><a href="#cb2-35" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb2-36"><a href="#cb2-36" aria-hidden="true" tabindex="-1"></a>}, <span class="at">x =</span> <span class="fu">list</span>(num_IVs,num_IVs), <span class="at">y =</span> <span class="fu">c</span>(<span class="st">"cis"</span>,<span class="st">"trans"</span>), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)</span>
<span id="cb2-37"><a href="#cb2-37" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb2-38"><a href="#cb2-38" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> p1)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-numivs" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-numivs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-numivs-1.png" id="fig-numivs" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-numivs-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;1
</figcaption>
</figure>
</div>
</div>
</div>
<p><strong>The number of proteins able to be instrumented with pQTLs across instrument sets</strong></p>
<p>Nearing 100% of O-link plasma proteins can be instrumented with either rare or common trans-IVs derived from WES data relative to &lt;75% when using GWAS derived pQTLs. Instrumenting proteins using rare cis-IVs from WES provides greater coverage of the plasma proteome than using common GWAS derived pQTLs, though common GWAS pQTLs provide better coverage than common WES derived pQTLs, suggesting cis-regulatory variants outside the coding region have significant effects on protein levels. Here, only instrument sets including &gt;=3 independent variants are considered.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb3"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb3-1"><a href="#cb3-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of proteins with significant causal estimates across instrument sets</span></span>
<span id="cb3-2"><a href="#cb3-2" aria-hidden="true" tabindex="-1"></a>summary_sig <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, summarise_results, <span class="at">p_thresh =</span> alpha, <span class="at">n_snps =</span> <span class="dv">3</span>)</span>
<span id="cb3-3"><a href="#cb3-3" aria-hidden="true" tabindex="-1"></a>summary_sig_cis <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, summarise_results, <span class="at">p_thresh =</span> alpha, <span class="at">n_snps =</span> <span class="dv">3</span>, <span class="at">cis_trans =</span> <span class="st">"cis"</span>)</span>
<span id="cb3-4"><a href="#cb3-4" aria-hidden="true" tabindex="-1"></a>summary_sig_trans <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, summarise_results, <span class="at">p_thresh =</span> alpha, <span class="at">n_snps =</span> <span class="dv">3</span>, <span class="at">cis_trans =</span> <span class="st">"trans"</span>)</span>
<span id="cb3-5"><a href="#cb3-5" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-6"><a href="#cb3-6" aria-hidden="true" tabindex="-1"></a><span class="co"># Number of proteins with instrument sets (of &gt;= n_snps) available</span></span>
<span id="cb3-7"><a href="#cb3-7" aria-hidden="true" tabindex="-1"></a>avail_IVs <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>,<span class="fu">lapply</span>(summary_sig, <span class="cf">function</span>(x){<span class="fu">apply</span>(x<span class="sc">$</span>min_p, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="cf">function</span>(y){<span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y))})<span class="sc">/</span><span class="fu">nrow</span>(x<span class="sc">$</span>min_p) <span class="sc">*</span> <span class="dv">100</span>}))</span>
<span id="cb3-8"><a href="#cb3-8" aria-hidden="true" tabindex="-1"></a>avail_IVs_cis <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>,<span class="fu">lapply</span>(summary_sig_cis, <span class="cf">function</span>(x){<span class="fu">apply</span>(x<span class="sc">$</span>min_p, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="cf">function</span>(y){<span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y))})<span class="sc">/</span><span class="fu">nrow</span>(x<span class="sc">$</span>min_p) <span class="sc">*</span> <span class="dv">100</span>}))</span>
<span id="cb3-9"><a href="#cb3-9" aria-hidden="true" tabindex="-1"></a>avail_IVs_trans <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>,<span class="fu">lapply</span>(summary_sig_trans, <span class="cf">function</span>(x){<span class="fu">apply</span>(x<span class="sc">$</span>min_p, <span class="at">MARGIN =</span> <span class="dv">2</span>, <span class="cf">function</span>(y){<span class="fu">sum</span>(<span class="sc">!</span><span class="fu">is.na</span>(y))})<span class="sc">/</span><span class="fu">nrow</span>(x<span class="sc">$</span>min_p) <span class="sc">*</span> <span class="dv">100</span>}))</span>
<span id="cb3-10"><a href="#cb3-10" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb3-11"><a href="#cb3-11" aria-hidden="true" tabindex="-1"></a>p2 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(<span class="at">FUN =</span> <span class="cf">function</span>(x,y){</span>
<span id="cb3-12"><a href="#cb3-12" aria-hidden="true" tabindex="-1"></a>  x[,<span class="fu">names</span>(variant_IVsets)] <span class="sc">|&gt;</span> </span>
<span id="cb3-13"><a href="#cb3-13" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span> </span>
<span id="cb3-14"><a href="#cb3-14" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"outcome"</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-15"><a href="#cb3-15" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="at">cols =</span> <span class="dv">2</span><span class="sc">:</span><span class="dv">8</span>) <span class="sc">|&gt;</span></span>
<span id="cb3-16"><a href="#cb3-16" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">factor</span>(variant_IVsets[name], <span class="at">levels =</span> variant_IVsets)) <span class="sc">|&gt;</span></span>
<span id="cb3-17"><a href="#cb3-17" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> label, <span class="at">y =</span> value, <span class="at">fill =</span> label)) <span class="sc">+</span></span>
<span id="cb3-18"><a href="#cb3-18" aria-hidden="true" tabindex="-1"></a>    <span class="fu">facet_grid</span>(.<span class="sc">~</span>outcome) <span class="sc">+</span></span>
<span id="cb3-19"><a href="#cb3-19" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_col</span>() <span class="sc">+</span></span>
<span id="cb3-20"><a href="#cb3-20" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">""</span>, <span class="at">values =</span> cols_variant_IVsets) <span class="sc">+</span></span>
<span id="cb3-21"><a href="#cb3-21" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(<span class="dv">0</span>,<span class="dv">100</span>)) <span class="sc">+</span></span>
<span id="cb3-22"><a href="#cb3-22" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb3-23"><a href="#cb3-23" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb3-24"><a href="#cb3-24" aria-hidden="true" tabindex="-1"></a>          <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb3-25"><a href="#cb3-25" aria-hidden="true" tabindex="-1"></a>          <span class="at">legend.position =</span> <span class="st">"none"</span>) <span class="sc">+</span></span>
<span id="cb3-26"><a href="#cb3-26" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"IV set"</span>,</span>
<span id="cb3-27"><a href="#cb3-27" aria-hidden="true" tabindex="-1"></a>         <span class="at">y =</span> <span class="st">"%"</span>, </span>
<span id="cb3-28"><a href="#cb3-28" aria-hidden="true" tabindex="-1"></a>         <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">"Percentage of plasma proteins with "</span>,y, <span class="st">"IV sets available (with n&gt;=3 instruments)"</span>))</span>
<span id="cb3-29"><a href="#cb3-29" aria-hidden="true" tabindex="-1"></a>}, <span class="at">x =</span> <span class="fu">list</span>(avail_IVs, avail_IVs_cis, avail_IVs_trans), <span class="at">y =</span> <span class="fu">c</span>(<span class="st">""</span>, <span class="st">"cis-"</span>, <span class="st">"trans-"</span>), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)</span>
<span id="cb3-30"><a href="#cb3-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb3-31"><a href="#cb3-31" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> p2)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-summary" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-summary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-summary-1.png" id="fig-summary" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-summary-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;2
</figcaption>
</figure>
</div>
</div>
</div>
<p><strong>Proportion of instrumented proteins with significant IVW causal effects detected across instrument sets</strong></p>
<p>Restricting to only those proteins that could be successfully instrumented (with n &gt;= 3 IVs) by all strategies shown below, a significant causal effect of protein levels on an outcome tend to be more often detected using common cis-IVs from GWAS than rare and ultra-rare cis-IVs.</p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb4"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb4-1"><a href="#cb4-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Calculate the proportion of significant exposures across variant sets</span></span>
<span id="cb4-2"><a href="#cb4-2" aria-hidden="true" tabindex="-1"></a><span class="co"># Keep proteins instrumentable by all variant-based IV sets</span></span>
<span id="cb4-3"><a href="#cb4-3" aria-hidden="true" tabindex="-1"></a>prop_sig_IVs <span class="ot">&lt;-</span> <span class="cf">function</span>(res_summary, p_thresh){</span>
<span id="cb4-4"><a href="#cb4-4" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-5"><a href="#cb4-5" aria-hidden="true" tabindex="-1"></a>  df_p <span class="ot">&lt;-</span> res_summary<span class="sc">$</span>min_p[,<span class="fu">names</span>(variant_IVsets_indep)] <span class="sc">|&gt;</span></span>
<span id="cb4-6"><a href="#cb4-6" aria-hidden="true" tabindex="-1"></a>    <span class="fu">na.exclude</span>() <span class="sc">|&gt;</span></span>
<span id="cb4-7"><a href="#cb4-7" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb4-8"><a href="#cb4-8" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"row"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-9"><a href="#cb4-9" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">sub</span>(<span class="st">"_.*"</span>,<span class="st">""</span>,row),</span>
<span id="cb4-10"><a href="#cb4-10" aria-hidden="true" tabindex="-1"></a>                  <span class="at">outcome =</span> <span class="fu">sub</span>(<span class="st">".*_"</span>,<span class="st">""</span>,row)) <span class="sc">|&gt;</span></span>
<span id="cb4-11"><a href="#cb4-11" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">names_to =</span> <span class="st">"IV_set"</span>, <span class="at">values_to =</span> <span class="st">"pval"</span>)</span>
<span id="cb4-12"><a href="#cb4-12" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-13"><a href="#cb4-13" aria-hidden="true" tabindex="-1"></a>  df_beta <span class="ot">&lt;-</span> res_summary<span class="sc">$</span>min_p_beta[,<span class="fu">names</span>(variant_IVsets_indep)] <span class="sc">|&gt;</span></span>
<span id="cb4-14"><a href="#cb4-14" aria-hidden="true" tabindex="-1"></a>    <span class="fu">na.exclude</span>() <span class="sc">|&gt;</span></span>
<span id="cb4-15"><a href="#cb4-15" aria-hidden="true" tabindex="-1"></a>    <span class="fu">as.data.frame</span>() <span class="sc">|&gt;</span></span>
<span id="cb4-16"><a href="#cb4-16" aria-hidden="true" tabindex="-1"></a>    tibble<span class="sc">::</span><span class="fu">rownames_to_column</span>(<span class="st">"row"</span>) <span class="sc">|&gt;</span></span>
<span id="cb4-17"><a href="#cb4-17" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">gene =</span> <span class="fu">sub</span>(<span class="st">"_.*"</span>,<span class="st">""</span>,row),</span>
<span id="cb4-18"><a href="#cb4-18" aria-hidden="true" tabindex="-1"></a>                  <span class="at">outcome =</span> <span class="fu">sub</span>(<span class="st">".*_"</span>,<span class="st">""</span>,row)) <span class="sc">|&gt;</span></span>
<span id="cb4-19"><a href="#cb4-19" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_longer</span>(<span class="dv">2</span><span class="sc">:</span><span class="dv">6</span>, <span class="at">names_to =</span> <span class="st">"IV_set"</span>, <span class="at">values_to =</span> <span class="st">"beta"</span>)</span>
<span id="cb4-20"><a href="#cb4-20" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-21"><a href="#cb4-21" aria-hidden="true" tabindex="-1"></a>  df <span class="ot">&lt;-</span> <span class="fu">left_join</span>(df_p, df_beta) <span class="sc">|&gt;</span></span>
<span id="cb4-22"><a href="#cb4-22" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">call =</span> <span class="fu">ifelse</span>(pval <span class="sc">&lt;=</span> p_thresh <span class="sc">&amp;</span> beta <span class="sc">&gt;</span> <span class="dv">0</span>, <span class="st">"sig_up"</span>,</span>
<span id="cb4-23"><a href="#cb4-23" aria-hidden="true" tabindex="-1"></a>                                <span class="fu">ifelse</span>(pval <span class="sc">&lt;=</span> p_thresh <span class="sc">&amp;</span> beta <span class="sc">&lt;</span> <span class="dv">0</span>, <span class="st">"sig_down"</span>, <span class="st">"ns"</span>)))</span>
<span id="cb4-24"><a href="#cb4-24" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-25"><a href="#cb4-25" aria-hidden="true" tabindex="-1"></a>  df_grouped <span class="ot">&lt;-</span> df <span class="sc">|&gt;</span> </span>
<span id="cb4-26"><a href="#cb4-26" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">group_by</span>(IV_set,call) <span class="sc">|&gt;</span></span>
<span id="cb4-27"><a href="#cb4-27" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">summarise</span>(<span class="at">n_genes =</span> <span class="fu">n</span>()) <span class="sc">|&gt;</span></span>
<span id="cb4-28"><a href="#cb4-28" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">prop =</span> n_genes<span class="sc">/</span><span class="fu">sum</span>(n_genes) <span class="sc">*</span> <span class="dv">100</span>,</span>
<span id="cb4-29"><a href="#cb4-29" aria-hidden="true" tabindex="-1"></a>                  <span class="at">outcome =</span> <span class="fu">unique</span>(df<span class="sc">$</span>outcome),</span>
<span id="cb4-30"><a href="#cb4-30" aria-hidden="true" tabindex="-1"></a>                  <span class="at">n_genes_tot =</span> <span class="fu">sum</span>(n_genes))</span>
<span id="cb4-31"><a href="#cb4-31" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb4-32"><a href="#cb4-32" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df_grouped)</span>
<span id="cb4-33"><a href="#cb4-33" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb4-34"><a href="#cb4-34" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-35"><a href="#cb4-35" aria-hidden="true" tabindex="-1"></a>sig_IVs <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig, prop_sig_IVs, <span class="at">p_thresh =</span> alpha))</span>
<span id="cb4-36"><a href="#cb4-36" aria-hidden="true" tabindex="-1"></a>sig_IVs_cis <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig_cis, prop_sig_IVs, <span class="at">p_thresh =</span> alpha))</span>
<span id="cb4-37"><a href="#cb4-37" aria-hidden="true" tabindex="-1"></a>sig_IVs_trans <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig_trans, prop_sig_IVs, <span class="at">p_thresh =</span> alpha))</span>
<span id="cb4-38"><a href="#cb4-38" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-39"><a href="#cb4-39" aria-hidden="true" tabindex="-1"></a>sig_IVs_0<span class="fl">.01</span> <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig, prop_sig_IVs, <span class="at">p_thresh =</span> <span class="fl">0.01</span>))</span>
<span id="cb4-40"><a href="#cb4-40" aria-hidden="true" tabindex="-1"></a>sig_IVs_0<span class="fl">.01</span>_cis <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig_cis, prop_sig_IVs, <span class="at">p_thresh =</span> <span class="fl">0.01</span>))</span>
<span id="cb4-41"><a href="#cb4-41" aria-hidden="true" tabindex="-1"></a>sig_IVs_0<span class="fl">.01</span>_trans <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>, <span class="fu">lapply</span>(summary_sig_trans, prop_sig_IVs, <span class="at">p_thresh =</span> <span class="fl">0.01</span>))</span>
<span id="cb4-42"><a href="#cb4-42" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-43"><a href="#cb4-43" aria-hidden="true" tabindex="-1"></a>p3 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(<span class="cf">function</span>(x,y){</span>
<span id="cb4-44"><a href="#cb4-44" aria-hidden="true" tabindex="-1"></a>  x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(call <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"sig_up"</span>,<span class="st">"sig_down"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-45"><a href="#cb4-45" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">factor</span>(variant_IVsets_indep[IV_set], <span class="at">levels =</span> variant_IVsets_indep)) <span class="sc">|&gt;</span></span>
<span id="cb4-46"><a href="#cb4-46" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> label, <span class="at">y =</span> prop, <span class="at">fill =</span> call)) <span class="sc">+</span></span>
<span id="cb4-47"><a href="#cb4-47" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(.<span class="sc">~</span>outcome) <span class="sc">+</span></span>
<span id="cb4-48"><a href="#cb4-48" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"stack"</span>, <span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb4-49"><a href="#cb4-49" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Direction of effect"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"steelblue"</span>,<span class="st">"tomato"</span>), <span class="at">label =</span> <span class="fu">c</span>(<span class="st">"down"</span>,<span class="st">"up"</span>)) <span class="sc">+</span></span>
<span id="cb4-50"><a href="#cb4-50" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb4-51"><a href="#cb4-51" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb4-52"><a href="#cb4-52" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb4-53"><a href="#cb4-53" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb4-54"><a href="#cb4-54" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">5</span>, <span class="at">r =</span> <span class="dv">5</span>, <span class="at">b =</span> <span class="dv">5</span>, <span class="at">l =</span> <span class="dv">30</span>)) <span class="sc">+</span></span>
<span id="cb4-55"><a href="#cb4-55" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"IV set"</span>,</span>
<span id="cb4-56"><a href="#cb4-56" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"%"</span>, </span>
<span id="cb4-57"><a href="#cb4-57" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">"Percentage of plasma protein causally associated with outcome using "</span>, y, <span class="st">"IVs (at P&lt;=0.01)"</span>))</span>
<span id="cb4-58"><a href="#cb4-58" aria-hidden="true" tabindex="-1"></a>}, <span class="at">x =</span> <span class="fu">list</span>(sig_IVs_0<span class="fl">.01</span>, sig_IVs_0<span class="fl">.01</span>_cis, sig_IVs_0<span class="fl">.01</span>_trans), <span class="at">y =</span> <span class="fu">c</span>(<span class="st">"all "</span>,<span class="st">"cis-"</span>,<span class="st">"trans-"</span>), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-59"><a href="#cb4-59" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-60"><a href="#cb4-60" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> p3)</span>
<span id="cb4-61"><a href="#cb4-61" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-62"><a href="#cb4-62" aria-hidden="true" tabindex="-1"></a>p4 <span class="ot">&lt;-</span> <span class="fu">mapply</span>(<span class="cf">function</span>(x,y){</span>
<span id="cb4-63"><a href="#cb4-63" aria-hidden="true" tabindex="-1"></a>  x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(call <span class="sc">%in%</span> <span class="fu">c</span>(<span class="st">"sig_up"</span>,<span class="st">"sig_down"</span>)) <span class="sc">|&gt;</span></span>
<span id="cb4-64"><a href="#cb4-64" aria-hidden="true" tabindex="-1"></a>  dplyr<span class="sc">::</span><span class="fu">mutate</span>(<span class="at">label =</span> <span class="fu">factor</span>(variant_IVsets_indep[IV_set], <span class="at">levels =</span> variant_IVsets_indep)) <span class="sc">|&gt;</span></span>
<span id="cb4-65"><a href="#cb4-65" aria-hidden="true" tabindex="-1"></a>  <span class="fu">ggplot</span>(<span class="fu">aes</span>(<span class="at">x =</span> label, <span class="at">y =</span> prop, <span class="at">fill =</span> call)) <span class="sc">+</span></span>
<span id="cb4-66"><a href="#cb4-66" aria-hidden="true" tabindex="-1"></a>  <span class="fu">facet_grid</span>(.<span class="sc">~</span>outcome) <span class="sc">+</span></span>
<span id="cb4-67"><a href="#cb4-67" aria-hidden="true" tabindex="-1"></a>  <span class="fu">geom_bar</span>(<span class="at">position =</span> <span class="st">"stack"</span>, <span class="at">stat =</span> <span class="st">"identity"</span>) <span class="sc">+</span></span>
<span id="cb4-68"><a href="#cb4-68" aria-hidden="true" tabindex="-1"></a>  <span class="fu">scale_fill_manual</span>(<span class="at">name =</span> <span class="st">"Direction of effect"</span>, <span class="at">values =</span> <span class="fu">c</span>(<span class="st">"steelblue"</span>,<span class="st">"tomato"</span>), <span class="at">label =</span> <span class="fu">c</span>(<span class="st">"down"</span>,<span class="st">"up"</span>)) <span class="sc">+</span></span>
<span id="cb4-69"><a href="#cb4-69" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb4-70"><a href="#cb4-70" aria-hidden="true" tabindex="-1"></a>  <span class="fu">theme</span>(<span class="at">axis.text.x =</span> <span class="fu">element_text</span>(<span class="at">angle =</span> <span class="dv">45</span>, <span class="at">hjust =</span> <span class="dv">1</span>),</span>
<span id="cb4-71"><a href="#cb4-71" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>),</span>
<span id="cb4-72"><a href="#cb4-72" aria-hidden="true" tabindex="-1"></a>        <span class="at">legend.position =</span> <span class="st">"bottom"</span>,</span>
<span id="cb4-73"><a href="#cb4-73" aria-hidden="true" tabindex="-1"></a>        <span class="at">plot.margin =</span> <span class="fu">margin</span>(<span class="at">t =</span> <span class="dv">5</span>, <span class="at">r =</span> <span class="dv">5</span>, <span class="at">b =</span> <span class="dv">5</span>, <span class="at">l =</span> <span class="dv">30</span>)) <span class="sc">+</span></span>
<span id="cb4-74"><a href="#cb4-74" aria-hidden="true" tabindex="-1"></a>  <span class="fu">labs</span>(<span class="at">x =</span> <span class="st">"IV set"</span>,</span>
<span id="cb4-75"><a href="#cb4-75" aria-hidden="true" tabindex="-1"></a>       <span class="at">y =</span> <span class="st">"%"</span>, </span>
<span id="cb4-76"><a href="#cb4-76" aria-hidden="true" tabindex="-1"></a>       <span class="at">title =</span> <span class="fu">paste0</span>(<span class="st">"Percentage of plasma protein causally associated with outcome using "</span>, y, <span class="st">"IVs (at P&lt;=1.7x10^-5)"</span>))</span>
<span id="cb4-77"><a href="#cb4-77" aria-hidden="true" tabindex="-1"></a>}, <span class="at">x =</span> <span class="fu">list</span>(sig_IVs, sig_IVs_cis, sig_IVs_trans), <span class="at">y =</span> <span class="fu">c</span>(<span class="st">"all "</span>,<span class="st">"cis-"</span>,<span class="st">"trans-"</span>), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>)</span>
<span id="cb4-78"><a href="#cb4-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb4-79"><a href="#cb4-79" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> p4)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-sigIVs-1" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-sigIVs-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-sigIVs-1.png" id="fig-sigIVs-1" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-sigIVs-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;3
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-sigIVs-2" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-sigIVs-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-sigIVs-2.png" id="fig-sigIVs-2" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-sigIVs-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;4
</figcaption>
</figure>
</div>
</div>
</div>
<p><strong>Comparison of IVW effect estimates across instrument sets</strong></p>
<div class="cell">
<details class="code-fold">
<summary>Code</summary>
<div class="sourceCode cell-code" id="cb5"><pre class="sourceCode r code-with-copy"><code class="sourceCode r"><span id="cb5-1"><a href="#cb5-1" aria-hidden="true" tabindex="-1"></a><span class="co"># Generate a table of IVW estimates for a pair of instrument sets</span></span>
<span id="cb5-2"><a href="#cb5-2" aria-hidden="true" tabindex="-1"></a>compare_estimates <span class="ot">&lt;-</span> <span class="cf">function</span>(results_MR, IVset1, IVset2, <span class="at">n_snps=</span><span class="dv">3</span>, <span class="at">cis_trans=</span><span class="st">"all"</span>){</span>
<span id="cb5-3"><a href="#cb5-3" aria-hidden="true" tabindex="-1"></a>  <span class="co"># Pull out IVW estimates and SE for instrument sets</span></span>
<span id="cb5-4"><a href="#cb5-4" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(cis_trans <span class="sc">==</span> <span class="st">"all"</span>){</span>
<span id="cb5-5"><a href="#cb5-5" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(results_MR, <span class="cf">function</span>(x){</span>
<span id="cb5-6"><a href="#cb5-6" aria-hidden="true" tabindex="-1"></a>      x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(method <span class="sc">==</span> <span class="st">"Inverse variance weighted"</span>, </span>
<span id="cb5-7"><a href="#cb5-7" aria-hidden="true" tabindex="-1"></a>                         cis_trans <span class="sc">==</span> <span class="st">"cis_trans"</span>,</span>
<span id="cb5-8"><a href="#cb5-8" aria-hidden="true" tabindex="-1"></a>                         IV_set <span class="sc">%in%</span> <span class="fu">c</span>(IVset1,IVset2))</span>
<span id="cb5-9"><a href="#cb5-9" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb5-10"><a href="#cb5-10" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb5-11"><a href="#cb5-11" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(cis_trans <span class="sc">==</span> <span class="st">"cis"</span>){</span>
<span id="cb5-12"><a href="#cb5-12" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(results_MR, <span class="cf">function</span>(x){</span>
<span id="cb5-13"><a href="#cb5-13" aria-hidden="true" tabindex="-1"></a>      x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(method <span class="sc">==</span> <span class="st">"Inverse variance weighted"</span>, </span>
<span id="cb5-14"><a href="#cb5-14" aria-hidden="true" tabindex="-1"></a>                         cis_trans <span class="sc">==</span> <span class="st">"cis"</span>,</span>
<span id="cb5-15"><a href="#cb5-15" aria-hidden="true" tabindex="-1"></a>                         IV_set <span class="sc">%in%</span> <span class="fu">c</span>(IVset1,IVset2))</span>
<span id="cb5-16"><a href="#cb5-16" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb5-17"><a href="#cb5-17" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb5-18"><a href="#cb5-18" aria-hidden="true" tabindex="-1"></a>  <span class="cf">if</span>(cis_trans <span class="sc">==</span> <span class="st">"trans"</span>){</span>
<span id="cb5-19"><a href="#cb5-19" aria-hidden="true" tabindex="-1"></a>    df <span class="ot">&lt;-</span> <span class="fu">lapply</span>(results_MR, <span class="cf">function</span>(x){</span>
<span id="cb5-20"><a href="#cb5-20" aria-hidden="true" tabindex="-1"></a>      x <span class="sc">|&gt;</span> dplyr<span class="sc">::</span><span class="fu">filter</span>(method <span class="sc">==</span> <span class="st">"Inverse variance weighted"</span>, </span>
<span id="cb5-21"><a href="#cb5-21" aria-hidden="true" tabindex="-1"></a>                         cis_trans <span class="sc">==</span> <span class="st">"trans"</span>,</span>
<span id="cb5-22"><a href="#cb5-22" aria-hidden="true" tabindex="-1"></a>                         IV_set <span class="sc">%in%</span> <span class="fu">c</span>(IVset1,IVset2))</span>
<span id="cb5-23"><a href="#cb5-23" aria-hidden="true" tabindex="-1"></a>    })</span>
<span id="cb5-24"><a href="#cb5-24" aria-hidden="true" tabindex="-1"></a>  }</span>
<span id="cb5-25"><a href="#cb5-25" aria-hidden="true" tabindex="-1"></a>  df_out_b <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>,df) <span class="sc">|&gt;</span> </span>
<span id="cb5-26"><a href="#cb5-26" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(nsnp <span class="sc">&gt;=</span> n_snps) <span class="sc">|&gt;</span></span>
<span id="cb5-27"><a href="#cb5-27" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"IV_set"</span>, <span class="at">values_from =</span> <span class="st">"b"</span>, <span class="at">id_cols =</span> <span class="st">"exposure"</span>) <span class="sc">|&gt;</span></span>
<span id="cb5-28"><a href="#cb5-28" aria-hidden="true" tabindex="-1"></a>    <span class="fu">na.exclude</span>() <span class="sc">|&gt;</span></span>
<span id="cb5-29"><a href="#cb5-29" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(exposure,IVset1,IVset2)</span>
<span id="cb5-30"><a href="#cb5-30" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-31"><a href="#cb5-31" aria-hidden="true" tabindex="-1"></a>  df_out_se <span class="ot">&lt;-</span> <span class="fu">do.call</span>(<span class="st">"rbind"</span>,df) <span class="sc">|&gt;</span> </span>
<span id="cb5-32"><a href="#cb5-32" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">filter</span>(nsnp <span class="sc">&gt;=</span> n_snps) <span class="sc">|&gt;</span></span>
<span id="cb5-33"><a href="#cb5-33" aria-hidden="true" tabindex="-1"></a>    tidyr<span class="sc">::</span><span class="fu">pivot_wider</span>(<span class="at">names_from =</span> <span class="st">"IV_set"</span>, <span class="at">values_from =</span> <span class="st">"se"</span>, <span class="at">id_cols =</span> <span class="st">"exposure"</span>) <span class="sc">|&gt;</span></span>
<span id="cb5-34"><a href="#cb5-34" aria-hidden="true" tabindex="-1"></a>    <span class="fu">na.exclude</span>() <span class="sc">|&gt;</span></span>
<span id="cb5-35"><a href="#cb5-35" aria-hidden="true" tabindex="-1"></a>    dplyr<span class="sc">::</span><span class="fu">relocate</span>(exposure,IVset1,IVset2)</span>
<span id="cb5-36"><a href="#cb5-36" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-37"><a href="#cb5-37" aria-hidden="true" tabindex="-1"></a>  df_out <span class="ot">&lt;-</span> dplyr<span class="sc">::</span><span class="fu">left_join</span>(df_out_b, df_out_se, <span class="at">by =</span> (<span class="st">"exposure"</span>), <span class="at">suffix =</span> <span class="fu">c</span>(<span class="st">".beta"</span>,<span class="st">".se"</span>))</span>
<span id="cb5-38"><a href="#cb5-38" aria-hidden="true" tabindex="-1"></a>  <span class="fu">return</span>(df_out)</span>
<span id="cb5-39"><a href="#cb5-39" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-40"><a href="#cb5-40" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-41"><a href="#cb5-41" aria-hidden="true" tabindex="-1"></a>plot_estimates <span class="ot">&lt;-</span> <span class="cf">function</span>(estimates, name){</span>
<span id="cb5-42"><a href="#cb5-42" aria-hidden="true" tabindex="-1"></a>  estimates <span class="ot">&lt;-</span> <span class="fu">as.data.frame</span>(estimates)</span>
<span id="cb5-43"><a href="#cb5-43" aria-hidden="true" tabindex="-1"></a>  col_names <span class="ot">&lt;-</span> <span class="fu">names</span>(estimates)</span>
<span id="cb5-44"><a href="#cb5-44" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-45"><a href="#cb5-45" aria-hidden="true" tabindex="-1"></a>  range_min <span class="ot">&lt;-</span> <span class="fu">min</span>(<span class="fu">min</span>(estimates[,<span class="dv">2</span>]), <span class="fu">min</span>(estimates[,<span class="dv">3</span>]))</span>
<span id="cb5-46"><a href="#cb5-46" aria-hidden="true" tabindex="-1"></a>  range_max <span class="ot">&lt;-</span> <span class="fu">max</span>(<span class="fu">max</span>(estimates[,<span class="dv">2</span>]), <span class="fu">max</span>(estimates[,<span class="dv">3</span>]))</span>
<span id="cb5-47"><a href="#cb5-47" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-48"><a href="#cb5-48" aria-hidden="true" tabindex="-1"></a>  lab_x <span class="ot">&lt;-</span> IVsets[<span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.beta"</span>,<span class="st">""</span>,col_names[<span class="dv">2</span>])]</span>
<span id="cb5-49"><a href="#cb5-49" aria-hidden="true" tabindex="-1"></a>  lab_y <span class="ot">&lt;-</span> IVsets[<span class="fu">sub</span>(<span class="st">"</span><span class="sc">\\</span><span class="st">.beta"</span>,<span class="st">""</span>,col_names[<span class="dv">3</span>])]</span>
<span id="cb5-50"><a href="#cb5-50" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-51"><a href="#cb5-51" aria-hidden="true" tabindex="-1"></a>  CI_x <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">LCI =</span> estimates[,<span class="dv">2</span>] <span class="sc">-</span> <span class="fl">1.96</span><span class="sc">*</span>estimates[,<span class="dv">4</span>], <span class="at">UCI =</span> estimates[,<span class="dv">2</span>] <span class="sc">+</span> <span class="fl">1.96</span><span class="sc">*</span>estimates[,<span class="dv">4</span>])</span>
<span id="cb5-52"><a href="#cb5-52" aria-hidden="true" tabindex="-1"></a>  CI_y <span class="ot">&lt;-</span> <span class="fu">data.frame</span>(<span class="at">LCI =</span> estimates[,<span class="dv">3</span>] <span class="sc">-</span> <span class="fl">1.96</span><span class="sc">*</span>estimates[,<span class="dv">5</span>], <span class="at">UCI =</span> estimates[,<span class="dv">3</span>] <span class="sc">+</span> <span class="fl">1.96</span><span class="sc">*</span>estimates[,<span class="dv">5</span>])</span>
<span id="cb5-53"><a href="#cb5-53" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-54"><a href="#cb5-54" aria-hidden="true" tabindex="-1"></a>  mod_lm <span class="ot">&lt;-</span> <span class="fu">lm</span>(estimates[,<span class="dv">3</span>] <span class="sc">~</span> estimates[,<span class="dv">2</span>])</span>
<span id="cb5-55"><a href="#cb5-55" aria-hidden="true" tabindex="-1"></a>  slope <span class="ot">&lt;-</span> <span class="fu">round</span>(<span class="fu">coefficients</span>(mod_lm)[<span class="dv">2</span>],<span class="dv">2</span>)</span>
<span id="cb5-56"><a href="#cb5-56" aria-hidden="true" tabindex="-1"></a>  </span>
<span id="cb5-57"><a href="#cb5-57" aria-hidden="true" tabindex="-1"></a>  estimates <span class="sc">|&gt;</span></span>
<span id="cb5-58"><a href="#cb5-58" aria-hidden="true" tabindex="-1"></a>    <span class="fu">ggplot</span>(<span class="fu">aes_string</span>(<span class="at">x =</span> col_names[<span class="dv">2</span>], <span class="at">y =</span> col_names[<span class="dv">3</span>])) <span class="sc">+</span></span>
<span id="cb5-59"><a href="#cb5-59" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_abline</span>(<span class="at">slope =</span> <span class="dv">1</span>, <span class="at">intercept =</span> <span class="dv">0</span>, <span class="at">colour =</span> <span class="st">"steelblue"</span>, <span class="at">lty =</span> <span class="dv">2</span>, <span class="at">lwd =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb5-60"><a href="#cb5-60" aria-hidden="true" tabindex="-1"></a>    <span class="co">#geom_errorbar(xmin = CI_x$LCI, xmax = CI_x$UCI, ymin = CI_y$LCI, ymax = CI_y$UCI) +</span></span>
<span id="cb5-61"><a href="#cb5-61" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_point</span>(<span class="at">size =</span> <span class="dv">2</span>, <span class="at">alpha =</span> <span class="fl">0.7</span>, <span class="at">shape =</span> <span class="dv">16</span>) <span class="sc">+</span></span>
<span id="cb5-62"><a href="#cb5-62" aria-hidden="true" tabindex="-1"></a>    <span class="fu">geom_smooth</span>(<span class="at">method =</span> <span class="st">"lm"</span>, <span class="at">se =</span> <span class="cn">FALSE</span>, <span class="at">colour =</span> <span class="st">"tomato"</span>, <span class="at">lwd =</span> <span class="fl">0.5</span>) <span class="sc">+</span></span>
<span id="cb5-63"><a href="#cb5-63" aria-hidden="true" tabindex="-1"></a>    <span class="fu">annotate</span>(<span class="at">geom =</span> <span class="st">"text"</span>, <span class="at">x =</span> range_min<span class="sc">/</span><span class="dv">2</span>, <span class="at">y =</span> range_max, <span class="at">label =</span> <span class="fu">paste0</span>(<span class="st">"Slope="</span>,slope), <span class="at">size =</span> <span class="dv">3</span>, <span class="at">colour =</span> <span class="st">"tomato"</span>) <span class="sc">+</span></span>
<span id="cb5-64"><a href="#cb5-64" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_x_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(range_min, range_max)) <span class="sc">+</span></span>
<span id="cb5-65"><a href="#cb5-65" aria-hidden="true" tabindex="-1"></a>    <span class="fu">scale_y_continuous</span>(<span class="at">limits =</span> <span class="fu">c</span>(range_min, range_max)) <span class="sc">+</span></span>
<span id="cb5-66"><a href="#cb5-66" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme_minimal</span>() <span class="sc">+</span></span>
<span id="cb5-67"><a href="#cb5-67" aria-hidden="true" tabindex="-1"></a>    <span class="fu">theme</span>(<span class="at">plot.title =</span> <span class="fu">element_text</span>(<span class="at">hjust =</span> <span class="fl">0.5</span>, <span class="at">size =</span> <span class="dv">10</span>)) <span class="sc">+</span></span>
<span id="cb5-68"><a href="#cb5-68" aria-hidden="true" tabindex="-1"></a>    <span class="fu">labs</span>(<span class="at">x=</span>lab_x, <span class="at">y=</span>lab_y, <span class="at">title =</span> name)</span>
<span id="cb5-69"><a href="#cb5-69" aria-hidden="true" tabindex="-1"></a>}</span>
<span id="cb5-70"><a href="#cb5-70" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-71"><a href="#cb5-71" aria-hidden="true" tabindex="-1"></a>comp_IVWs_common <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="at">FUN =</span> compare_estimates, </span>
<span id="cb5-72"><a href="#cb5-72" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset1 =</span> <span class="st">"dhindsa_exwas_variant_common"</span>,</span>
<span id="cb5-73"><a href="#cb5-73" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset2 =</span> <span class="st">"sun_gwas_variant_common_clumped"</span>,</span>
<span id="cb5-74"><a href="#cb5-74" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_snps =</span> <span class="dv">3</span>)</span>
<span id="cb5-75"><a href="#cb5-75" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-76"><a href="#cb5-76" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> <span class="fu">mapply</span>(plot_estimates, <span class="at">estimates =</span> comp_IVWs_common, <span class="at">name =</span> <span class="fu">names</span>(comp_IVWs_common), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>), <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb5-77"><a href="#cb5-77" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top =</span> <span class="st">"Comparison of IVW estimates from GWAS common and WES common pQTL instrumentation strategies"</span>)</span>
<span id="cb5-78"><a href="#cb5-78" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-79"><a href="#cb5-79" aria-hidden="true" tabindex="-1"></a>comp_IVWs <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="at">FUN =</span> compare_estimates, </span>
<span id="cb5-80"><a href="#cb5-80" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset1 =</span> <span class="st">"dhindsa_exwas_variant_common"</span>,</span>
<span id="cb5-81"><a href="#cb5-81" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset2 =</span> <span class="st">"dhindsa_exwas_variant_rare_filt"</span>,</span>
<span id="cb5-82"><a href="#cb5-82" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_snps =</span> <span class="dv">3</span>,</span>
<span id="cb5-83"><a href="#cb5-83" aria-hidden="true" tabindex="-1"></a>                    <span class="at">cis_trans =</span> <span class="st">"cis"</span>)</span>
<span id="cb5-84"><a href="#cb5-84" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-85"><a href="#cb5-85" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> <span class="fu">mapply</span>(plot_estimates, <span class="at">estimates =</span> comp_IVWs, <span class="at">name =</span> <span class="fu">names</span>(comp_IVWs), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>), <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb5-86"><a href="#cb5-86" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top =</span> <span class="st">"Comparison of IVW estimates from WES rare and common cis-pQTL instrumentation strategies"</span>)</span>
<span id="cb5-87"><a href="#cb5-87" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-88"><a href="#cb5-88" aria-hidden="true" tabindex="-1"></a>comp_IVWs_2 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="at">FUN =</span> compare_estimates, </span>
<span id="cb5-89"><a href="#cb5-89" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset1 =</span> <span class="st">"sun_gwas_variant_common_clumped"</span>,</span>
<span id="cb5-90"><a href="#cb5-90" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset2 =</span> <span class="st">"dhindsa_exwas_variant_rare_filt"</span>,</span>
<span id="cb5-91"><a href="#cb5-91" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_snps =</span> <span class="dv">3</span>,</span>
<span id="cb5-92"><a href="#cb5-92" aria-hidden="true" tabindex="-1"></a>                    <span class="at">cis_trans =</span> <span class="st">"cis"</span>)</span>
<span id="cb5-93"><a href="#cb5-93" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-94"><a href="#cb5-94" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> <span class="fu">mapply</span>(plot_estimates, <span class="at">estimates =</span> comp_IVWs_2, <span class="at">name =</span> <span class="fu">names</span>(comp_IVWs_2), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>), <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb5-95"><a href="#cb5-95" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top =</span> <span class="st">"Comparison of IVW estimates from WES rare and GWAS common cis-pQTL instrumentation strategies"</span>)</span>
<span id="cb5-96"><a href="#cb5-96" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-97"><a href="#cb5-97" aria-hidden="true" tabindex="-1"></a>comp_IVWs_3 <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="at">FUN =</span> compare_estimates, </span>
<span id="cb5-98"><a href="#cb5-98" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset1 =</span> <span class="st">"dhindsa_exwas_variant_common"</span>,</span>
<span id="cb5-99"><a href="#cb5-99" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset2 =</span> <span class="st">"dhindsa_exwas_variant_rare_filt"</span>,</span>
<span id="cb5-100"><a href="#cb5-100" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_snps =</span> <span class="dv">3</span>,</span>
<span id="cb5-101"><a href="#cb5-101" aria-hidden="true" tabindex="-1"></a>                    <span class="at">cis_trans =</span> <span class="st">"trans"</span>)</span>
<span id="cb5-102"><a href="#cb5-102" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-103"><a href="#cb5-103" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> <span class="fu">mapply</span>(plot_estimates, <span class="at">estimates =</span> comp_IVWs_3, <span class="at">name =</span> <span class="fu">names</span>(comp_IVWs_3), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>), <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb5-104"><a href="#cb5-104" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top =</span> <span class="st">"Comparison of IVW estimates from WES rare and common trans-pQTL instrumentation strategies"</span>)</span>
<span id="cb5-105"><a href="#cb5-105" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-106"><a href="#cb5-106" aria-hidden="true" tabindex="-1"></a>comp_IVWs_rare <span class="ot">&lt;-</span> <span class="fu">lapply</span>(res_MR, <span class="at">FUN =</span> compare_estimates, </span>
<span id="cb5-107"><a href="#cb5-107" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset1 =</span> <span class="st">"dhindsa_exwas_variant_rare"</span>,</span>
<span id="cb5-108"><a href="#cb5-108" aria-hidden="true" tabindex="-1"></a>                    <span class="at">IVset2 =</span> <span class="st">"dhindsa_exwas_variant_ultrarare"</span>,</span>
<span id="cb5-109"><a href="#cb5-109" aria-hidden="true" tabindex="-1"></a>                    <span class="at">n_snps =</span> <span class="dv">3</span>)</span>
<span id="cb5-110"><a href="#cb5-110" aria-hidden="true" tabindex="-1"></a></span>
<span id="cb5-111"><a href="#cb5-111" aria-hidden="true" tabindex="-1"></a>gridExtra<span class="sc">::</span><span class="fu">grid.arrange</span>(<span class="at">grobs =</span> <span class="fu">mapply</span>(plot_estimates, <span class="at">estimates =</span> comp_IVWs_rare, <span class="at">name =</span> <span class="fu">names</span>(comp_IVWs_rare), <span class="at">SIMPLIFY =</span> <span class="cn">FALSE</span>), <span class="at">nrow =</span> <span class="dv">2</span>,</span>
<span id="cb5-112"><a href="#cb5-112" aria-hidden="true" tabindex="-1"></a>                        <span class="at">top =</span> <span class="st">"Comparison of IVW estimates from WES rare and ultrarare pQTL instrumentation strategies"</span>)</span></code><button title="Copy to Clipboard" class="code-copy-button"><i class="bi"></i></button></pre></div>
</details>
<div class="cell-output-display">
<div id="fig-compIVs-1" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-compIVs-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-compIVs-1.png" id="fig-compIVs-1" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-compIVs-1-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;5
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-compIVs-2" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-compIVs-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-compIVs-2.png" id="fig-compIVs-2" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-compIVs-2-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;6
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-compIVs-3" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-compIVs-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-compIVs-3.png" id="fig-compIVs-3" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-compIVs-3-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;7
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-compIVs-4" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-compIVs-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-compIVs-4.png" id="fig-compIVs-4" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-compIVs-4-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;8
</figcaption>
</figure>
</div>
</div>
<div class="cell-output-display">
<div id="fig-compIVs-5" class="quarto-float quarto-figure quarto-figure-center anchored" width="1344">
<figure class="quarto-float quarto-float-fig figure">
<div aria-describedby="fig-compIVs-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
<img src="rarevcommon_molecularMR_files/figure-html/fig-compIVs-5.png" id="fig-compIVs-5" class="img-fluid figure-img" width="1344">
</div>
<figcaption class="quarto-float-caption-bottom quarto-float-caption quarto-float-fig quarto-uncaptioned" id="fig-compIVs-5-caption-0ceaefa1-69ba-4598-a22c-09a6ac19f8ca">
Figure&nbsp;9
</figcaption>
</figure>
</div>
</div>
</div>
</section>
<section id="next-steps" class="level2">
<h2 class="anchored" data-anchor-id="next-steps">Next steps</h2>
<ul>
<li><p>Determine which IVW estimates exhibit heterogenity between IV sets</p></li>
<li><p>Look into examples where:</p>
<ul>
<li>common variants give highly significant causal estimates and rare variants don’t and vice versa</li>
<li>differences between GWAS (regulatory) and WES (coding) common variant IVW estimates</li>
<li>differences between cis and trans instrumentation</li>
</ul></li>
<li><p>Some case studies with a more thorough interrogation of where the instruments are sitting and the biology they are instrumenting</p></li>
<li><p>Look at estimates from the gene-based masks</p></li>
<li><p>Pleiotropy:</p>
<ul>
<li>Compare the pleiotropy profiles of rare and common (and cis and trans) instruments</li>
</ul></li>
<li><p>Rare regulatory variants (required WGS data)</p></li>
</ul>
</section>

</main>
<!-- /main column -->
<script id="quarto-html-after-body" type="application/javascript">
window.document.addEventListener("DOMContentLoaded", function (event) {
  const toggleBodyColorMode = (bsSheetEl) => {
    const mode = bsSheetEl.getAttribute("data-mode");
    const bodyEl = window.document.querySelector("body");
    if (mode === "dark") {
      bodyEl.classList.add("quarto-dark");
      bodyEl.classList.remove("quarto-light");
    } else {
      bodyEl.classList.add("quarto-light");
      bodyEl.classList.remove("quarto-dark");
    }
  }
  const toggleBodyColorPrimary = () => {
    const bsSheetEl = window.document.querySelector("link#quarto-bootstrap");
    if (bsSheetEl) {
      toggleBodyColorMode(bsSheetEl);
    }
  }
  toggleBodyColorPrimary();  
  const icon = "";
  const anchorJS = new window.AnchorJS();
  anchorJS.options = {
    placement: 'right',
    icon: icon
  };
  anchorJS.add('.anchored');
  const isCodeAnnotation = (el) => {
    for (const clz of el.classList) {
      if (clz.startsWith('code-annotation-')) {                     
        return true;
      }
    }
    return false;
  }
  const onCopySuccess = function(e) {
    // button target
    const button = e.trigger;
    // don't keep focus
    button.blur();
    // flash "checked"
    button.classList.add('code-copy-button-checked');
    var currentTitle = button.getAttribute("title");
    button.setAttribute("title", "Copied!");
    let tooltip;
    if (window.bootstrap) {
      button.setAttribute("data-bs-toggle", "tooltip");
      button.setAttribute("data-bs-placement", "left");
      button.setAttribute("data-bs-title", "Copied!");
      tooltip = new bootstrap.Tooltip(button, 
        { trigger: "manual", 
          customClass: "code-copy-button-tooltip",
          offset: [0, -8]});
      tooltip.show();    
    }
    setTimeout(function() {
      if (tooltip) {
        tooltip.hide();
        button.removeAttribute("data-bs-title");
        button.removeAttribute("data-bs-toggle");
        button.removeAttribute("data-bs-placement");
      }
      button.setAttribute("title", currentTitle);
      button.classList.remove('code-copy-button-checked');
    }, 1000);
    // clear code selection
    e.clearSelection();
  }
  const getTextToCopy = function(trigger) {
      const codeEl = trigger.previousElementSibling.cloneNode(true);
      for (const childEl of codeEl.children) {
        if (isCodeAnnotation(childEl)) {
          childEl.remove();
        }
      }
      return codeEl.innerText;
  }
  const clipboard = new window.ClipboardJS('.code-copy-button:not([data-in-quarto-modal])', {
    text: getTextToCopy
  });
  clipboard.on('success', onCopySuccess);
  if (window.document.getElementById('quarto-embedded-source-code-modal')) {
    const clipboardModal = new window.ClipboardJS('.code-copy-button[data-in-quarto-modal]', {
      text: getTextToCopy,
      container: window.document.getElementById('quarto-embedded-source-code-modal')
    });
    clipboardModal.on('success', onCopySuccess);
  }
    var localhostRegex = new RegExp(/^(?:http|https):\/\/localhost\:?[0-9]*\//);
    var mailtoRegex = new RegExp(/^mailto:/);
      var filterRegex = new RegExp('/' + window.location.host + '/');
    var isInternal = (href) => {
        return filterRegex.test(href) || localhostRegex.test(href) || mailtoRegex.test(href);
    }
    // Inspect non-navigation links and adorn them if external
 	var links = window.document.querySelectorAll('a[href]:not(.nav-link):not(.navbar-brand):not(.toc-action):not(.sidebar-link):not(.sidebar-item-toggle):not(.pagination-link):not(.no-external):not([aria-hidden]):not(.dropdown-item):not(.quarto-navigation-tool):not(.about-link)');
    for (var i=0; i<links.length; i++) {
      const link = links[i];
      if (!isInternal(link.href)) {
        // undo the damage that might have been done by quarto-nav.js in the case of
        // links that we want to consider external
        if (link.dataset.originalHref !== undefined) {
          link.href = link.dataset.originalHref;
        }
      }
    }
  function tippyHover(el, contentFn, onTriggerFn, onUntriggerFn) {
    const config = {
      allowHTML: true,
      maxWidth: 500,
      delay: 100,
      arrow: false,
      appendTo: function(el) {
          return el.parentElement;
      },
      interactive: true,
      interactiveBorder: 10,
      theme: 'quarto',
      placement: 'bottom-start',
    };
    if (contentFn) {
      config.content = contentFn;
    }
    if (onTriggerFn) {
      config.onTrigger = onTriggerFn;
    }
    if (onUntriggerFn) {
      config.onUntrigger = onUntriggerFn;
    }
    window.tippy(el, config); 
  }
  const noterefs = window.document.querySelectorAll('a[role="doc-noteref"]');
  for (var i=0; i<noterefs.length; i++) {
    const ref = noterefs[i];
    tippyHover(ref, function() {
      // use id or data attribute instead here
      let href = ref.getAttribute('data-footnote-href') || ref.getAttribute('href');
      try { href = new URL(href).hash; } catch {}
      const id = href.replace(/^#\/?/, "");
      const note = window.document.getElementById(id);
      if (note) {
        return note.innerHTML;
      } else {
        return "";
      }
    });
  }
  const xrefs = window.document.querySelectorAll('a.quarto-xref');
  const processXRef = (id, note) => {
    // Strip column container classes
    const stripColumnClz = (el) => {
      el.classList.remove("page-full", "page-columns");
      if (el.children) {
        for (const child of el.children) {
          stripColumnClz(child);
        }
      }
    }
    stripColumnClz(note)
    if (id === null || id.startsWith('sec-')) {
      // Special case sections, only their first couple elements
      const container = document.createElement("div");
      if (note.children && note.children.length > 2) {
        container.appendChild(note.children[0].cloneNode(true));
        for (let i = 1; i < note.children.length; i++) {
          const child = note.children[i];
          if (child.tagName === "P" && child.innerText === "") {
            continue;
          } else {
            container.appendChild(child.cloneNode(true));
            break;
          }
        }
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(container);
        }
        return container.innerHTML
      } else {
        if (window.Quarto?.typesetMath) {
          window.Quarto.typesetMath(note);
        }
        return note.innerHTML;
      }
    } else {
      // Remove any anchor links if they are present
      const anchorLink = note.querySelector('a.anchorjs-link');
      if (anchorLink) {
        anchorLink.remove();
      }
      if (window.Quarto?.typesetMath) {
        window.Quarto.typesetMath(note);
      }
      if (note.classList.contains("callout")) {
        return note.outerHTML;
      } else {
        return note.innerHTML;
      }
    }
  }
  for (var i=0; i<xrefs.length; i++) {
    const xref = xrefs[i];
    tippyHover(xref, undefined, function(instance) {
      instance.disable();
      let url = xref.getAttribute('href');
      let hash = undefined; 
      if (url.startsWith('#')) {
        hash = url;
      } else {
        try { hash = new URL(url).hash; } catch {}
      }
      if (hash) {
        const id = hash.replace(/^#\/?/, "");
        const note = window.document.getElementById(id);
        if (note !== null) {
          try {
            const html = processXRef(id, note.cloneNode(true));
            instance.setContent(html);
          } finally {
            instance.enable();
            instance.show();
          }
        } else {
          // See if we can fetch this
          fetch(url.split('#')[0])
          .then(res => res.text())
          .then(html => {
            const parser = new DOMParser();
            const htmlDoc = parser.parseFromString(html, "text/html");
            const note = htmlDoc.getElementById(id);
            if (note !== null) {
              const html = processXRef(id, note);
              instance.setContent(html);
            } 
          }).finally(() => {
            instance.enable();
            instance.show();
          });
        }
      } else {
        // See if we can fetch a full url (with no hash to target)
        // This is a special case and we should probably do some content thinning / targeting
        fetch(url)
        .then(res => res.text())
        .then(html => {
          const parser = new DOMParser();
          const htmlDoc = parser.parseFromString(html, "text/html");
          const note = htmlDoc.querySelector('main.content');
          if (note !== null) {
            // This should only happen for chapter cross references
            // (since there is no id in the URL)
            // remove the first header
            if (note.children.length > 0 && note.children[0].tagName === "HEADER") {
              note.children[0].remove();
            }
            const html = processXRef(null, note);
            instance.setContent(html);
          } 
        }).finally(() => {
          instance.enable();
          instance.show();
        });
      }
    }, function(instance) {
    });
  }
      let selectedAnnoteEl;
      const selectorForAnnotation = ( cell, annotation) => {
        let cellAttr = 'data-code-cell="' + cell + '"';
        let lineAttr = 'data-code-annotation="' +  annotation + '"';
        const selector = 'span[' + cellAttr + '][' + lineAttr + ']';
        return selector;
      }
      const selectCodeLines = (annoteEl) => {
        const doc = window.document;
        const targetCell = annoteEl.getAttribute("data-target-cell");
        const targetAnnotation = annoteEl.getAttribute("data-target-annotation");
        const annoteSpan = window.document.querySelector(selectorForAnnotation(targetCell, targetAnnotation));
        const lines = annoteSpan.getAttribute("data-code-lines").split(",");
        const lineIds = lines.map((line) => {
          return targetCell + "-" + line;
        })
        let top = null;
        let height = null;
        let parent = null;
        if (lineIds.length > 0) {
            //compute the position of the single el (top and bottom and make a div)
            const el = window.document.getElementById(lineIds[0]);
            top = el.offsetTop;
            height = el.offsetHeight;
            parent = el.parentElement.parentElement;
          if (lineIds.length > 1) {
            const lastEl = window.document.getElementById(lineIds[lineIds.length - 1]);
            const bottom = lastEl.offsetTop + lastEl.offsetHeight;
            height = bottom - top;
          }
          if (top !== null && height !== null && parent !== null) {
            // cook up a div (if necessary) and position it 
            let div = window.document.getElementById("code-annotation-line-highlight");
            if (div === null) {
              div = window.document.createElement("div");
              div.setAttribute("id", "code-annotation-line-highlight");
              div.style.position = 'absolute';
              parent.appendChild(div);
            }
            div.style.top = top - 2 + "px";
            div.style.height = height + 4 + "px";
            div.style.left = 0;
            let gutterDiv = window.document.getElementById("code-annotation-line-highlight-gutter");
            if (gutterDiv === null) {
              gutterDiv = window.document.createElement("div");
              gutterDiv.setAttribute("id", "code-annotation-line-highlight-gutter");
              gutterDiv.style.position = 'absolute';
              const codeCell = window.document.getElementById(targetCell);
              const gutter = codeCell.querySelector('.code-annotation-gutter');
              gutter.appendChild(gutterDiv);
            }
            gutterDiv.style.top = top - 2 + "px";
            gutterDiv.style.height = height + 4 + "px";
          }
          selectedAnnoteEl = annoteEl;
        }
      };
      const unselectCodeLines = () => {
        const elementsIds = ["code-annotation-line-highlight", "code-annotation-line-highlight-gutter"];
        elementsIds.forEach((elId) => {
          const div = window.document.getElementById(elId);
          if (div) {
            div.remove();
          }
        });
        selectedAnnoteEl = undefined;
      };
        // Handle positioning of the toggle
    window.addEventListener(
      "resize",
      throttle(() => {
        elRect = undefined;
        if (selectedAnnoteEl) {
          selectCodeLines(selectedAnnoteEl);
        }
      }, 10)
    );
    function throttle(fn, ms) {
    let throttle = false;
    let timer;
      return (...args) => {
        if(!throttle) { // first call gets through
            fn.apply(this, args);
            throttle = true;
        } else { // all the others get throttled
            if(timer) clearTimeout(timer); // cancel #2
            timer = setTimeout(() => {
              fn.apply(this, args);
              timer = throttle = false;
            }, ms);
        }
      };
    }
      // Attach click handler to the DT
      const annoteDls = window.document.querySelectorAll('dt[data-target-cell]');
      for (const annoteDlNode of annoteDls) {
        annoteDlNode.addEventListener('click', (event) => {
          const clickedEl = event.target;
          if (clickedEl !== selectedAnnoteEl) {
            unselectCodeLines();
            const activeEl = window.document.querySelector('dt[data-target-cell].code-annotation-active');
            if (activeEl) {
              activeEl.classList.remove('code-annotation-active');
            }
            selectCodeLines(clickedEl);
            clickedEl.classList.add('code-annotation-active');
          } else {
            // Unselect the line
            unselectCodeLines();
            clickedEl.classList.remove('code-annotation-active');
          }
        });
      }
  const findCites = (el) => {
    const parentEl = el.parentElement;
    if (parentEl) {
      const cites = parentEl.dataset.cites;
      if (cites) {
        return {
          el,
          cites: cites.split(' ')
        };
      } else {
        return findCites(el.parentElement)
      }
    } else {
      return undefined;
    }
  };
  var bibliorefs = window.document.querySelectorAll('a[role="doc-biblioref"]');
  for (var i=0; i<bibliorefs.length; i++) {
    const ref = bibliorefs[i];
    const citeInfo = findCites(ref);
    if (citeInfo) {
      tippyHover(citeInfo.el, function() {
        var popup = window.document.createElement('div');
        citeInfo.cites.forEach(function(cite) {
          var citeDiv = window.document.createElement('div');
          citeDiv.classList.add('hanging-indent');
          citeDiv.classList.add('csl-entry');
          var biblioDiv = window.document.getElementById('ref-' + cite);
          if (biblioDiv) {
            citeDiv.innerHTML = biblioDiv.innerHTML;
          }
          popup.appendChild(citeDiv);
        });
        return popup.innerHTML;
      });
    }
  }
});
</script>
</div> <!-- /content -->




</body></html>