---
title: "Comparing common and rare single variant and gene aggregate instrumentation strategies for MR"
author: "Aimee Hanson"
date: "`r format(Sys.time(), '%d %B %Y')`"
format: 
  html:
    toc: true
    toc-depth: 4
    toc-location: left
---

```{r}
#| label: setup

library(dotenv)
library(here)
library(dplyr)
library(ggplot2)

dotenv::load_dot_env(file = here("config.env"))
data_dir <- Sys.getenv("data_dir")
res_dir <- Sys.getenv("results_dir")
```

## Introduction

Classically, Mendelian Randomisation (MR) methods utilising trait-associated genetic variants from GWAS studies have employed common polymorphisms (population MAF \> 1%) targeted by genotyping chips, or reliably imputed from reference populations, to instrument modifiable exposures. Hoewever, common variants tested within GWAS typically explain a very small fraction of the variability in a measured complex trait. Though easily estimated by MR methods, the physiological/clinical relevance of subtle, genetically proxied perturbations in an exposure varaible on an outcome of interest may be unreliable to interpret in the context of the magnitude of observed trait variation in a population. The vast majority of common variant trait associations also tend to fall outside functionally annotated genomic regions, potentially tagging regulatory sites with non-specific roles in a diversity of biological processes. These pleiotropic varaint profiles can introduce heterogeneity into causal effect estimates, and violate the exclusion restriction assumption.

Rare variants, which typically show large biological effects (e.g. through abolising protein expression) provide a means of more unambiguously instrumenting relevant molecular processes that underly variation in measured exposures. Significant rare variant association identified by GWAS are also more likely to represent causal variation, rarther than tagging haplotypes of linked genomic sites... CONTINUE

\[Finish this rambling\] Comparison of causal estimates derived using differing methods of genetically instrumenting modifiable exposures may enhance the interpretation of the biological mechanisms underlying exposure-outcome relationships.

### Instruments

Twelve sets of instruments for each exposure have been extracted from across three sources (DeepRVAT gene impairment scores, Genebass whole exome single variants and aggregate burden masks and UKB common variant GWAS summary statistics):

**Common GWAS**\
Associated common variants from UKB GWAS (\>1% MAF)

**Genebass (variants)**\
Common exome-wide (\>5% MAF, LD clumped)\
Low-frequency exome-wide (1-5% MAF, LD clumped)\
Rare exome-wide (0-1% MAF, both unfiltered and filtered to the top hit per gene)\
Ultra-rare exome-wide (0-0.1% MAF, both unfiltered and filtered to the top hit per gene)

**Genebass (masks)**\
pLOF burden mask\
missense/low-confidence burden mask\
pLOF/missense/low-confidence burden mask\
synonymous burden mask

**DeepRVAT (impairment scores)**\
gene impairment score (\<0.1% MAF)

## Data import

Read in harmonised instrument summary statistics and filter out variants with MAF\<1x10\^-5.

```{r}
#| label: import harmonised studies

filelist <- list.files(path = file.path(data_dir, "harmonised"), pattern = ".rda")

# Exposure - outcome study pairs
study_pairs <- sub("^.*_.*_(.*_.*).rda","\\1",filelist) |> unique()

harmonised_studies <- as.list(rep(NA, length(study_pairs)))
names(harmonised_studies) <- study_pairs

for (pair in study_pairs){
  files <- file.path(data_dir,"harmonised", 
                  grep(pair, list.files(path = file.path(data_dir, "harmonised")), value = T))
  file_source <- sub(paste0("_",pair,".rda"),"",basename(files))
  names(files) <- file_source
  
  for (file in file_source){
    env <- new.env()
    
    if(file == "deeprvat_genescore"){
      load(files[file], envir = env)
      obj1 <- as.list(env)
      names(obj1) <- "deeprvat_genescore"
    }
    
    else if(file == "genebass_mask"){
      load(files[file], envir = env)
      obj <- as.list(env)
      obj2 <- do.call(c, obj)
      names(obj2) <- paste0("genebass_",names(obj$dat_harmonised))
    }
    
    else if(file == "genebass_variant"){
      load(files[file], envir = env)
      obj <- as.list(env)
      obj3 <- do.call(c, obj)
      names(obj3) <- paste0("genebass_",names(obj$dat_harmonised))
    }
    
    else if(file == "opengwas_variant"){
      load(files[file], envir = env)
      obj4 <- as.list(env)
      names(obj4) <- paste0("opengwas_common")
    }
  }
  
  out_studies <- c(obj1,obj2,obj3,obj4)
  
  # Add column names to empty dataframes
  col_names <- lapply(out_studies, colnames) |> unlist() |> unique()
  
  if(any(lapply(out_studies,nrow) == 0)){
    empty_df <- which(lapply(out_studies,nrow) == 0)
    for(df in empty_df){
      out_studies[[df]] <- data.frame(matrix(data = NA, nrow = 1, ncol = length(col_names), dimnames = list(NULL,col_names)))}
  }
  
  # Remove instruments with MAF<1x10^-5 (some present in genebass studies)
  out_studies <- lapply(out_studies, function(x){
    if(!is.na(x$eaf.exposure[1])){
      x |> dplyr::filter(eaf.exposure > 1*10^-5 & 1-eaf.exposure > 1*10^-5)
    }else{
      x
    }
  })
  
  harmonised_studies[[pair]] <- out_studies 
}

# Annotations for study sets
instrument_type <- c("genescore",rep("mask",4), rep("variant",7))
names(instrument_type) <- names(harmonised_studies[[1]])

instrument_class <- c("genescore","missense|LC","pLOF","pLOF|missense|LC","synonymous","exome:common","exome:lowfreq",
                      "exome:rare","exome:ultrarare","exome:rare","exome:ultrarare","genome:common")
names(instrument_class) <- names(harmonised_studies[[1]])

rm(obj,obj1,obj2,obj3,obj4,out_studies)

```

### Check instrument selection

Single variant instruments only:

```{r}
#| label: check instruments
#| message: FALSE
#| fig-width: 10
#| fig-height: 12

# Check number of instruments, minor allele frequency and average beta values across instrument sets
instrument_MAFs <- list()

for(i in 1:length(harmonised_studies)){
  MAFs <- do.call("rbind", lapply(harmonised_studies[[i]], function(x){
    x |> 
      mutate(maf = ifelse(eaf.exposure > 0.5, 1-eaf.exposure, eaf.exposure)) |>
      group_by(exposure,outcome) |>
      dplyr::summarise(
        n_instruments = n(),
        mean_maf = mean(maf),
        mean_beta = mean(abs(beta.exposure)))
  }))
  MAFs$pairs <- names(harmonised_studies)[i]
  MAFs$study <- names(harmonised_studies[[i]])
  MAFs$instrument_type <- instrument_type[MAFs$study]
  MAFs$instrument_class <- instrument_class[MAFs$study]
  
  instrument_MAFs <- c(instrument_MAFs, list(MAFs))
}

instrument_MAFs <- do.call("rbind", instrument_MAFs) |>
  mutate(study = factor(study, levels = c("opengwas_common","genebass_common",
                                         "genebass_lowfreq",
                                         "genebass_rare","genebass_rare_filt",
                                         "genebass_ultrarare","genebass_ultrarare_filt",
                                         "genebass_synonymous","genebass_missense|LC","genebass_pLoF|missense|LC","genebass_pLoF",
                                         "deeprvat_genescore")))

## Plot average instrument MAF

p_MAF <- ggplot(instrument_MAFs |> filter(instrument_type == "variant"),
       aes(x = study, y = mean_maf, size = n_instruments)) +
  geom_point() + 
  facet_wrap(.~pairs, nrow = 2) +
  geom_hline(yintercept = 0.01, lty = 2) +
  geom_hline(yintercept = 0.05, lty = 2) +
  geom_hline(yintercept = 0.001, lty = 2) +
  scale_y_log10() +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(y = "Mean instrument MAF", x = "Source study", 
       size = "n (instruments)")

## Plot average instrument beta

p_beta <- ggplot(instrument_MAFs |> filter(instrument_type == "variant"),
       aes(x = study, y = mean_beta, size = n_instruments)) +
  geom_point() + 
  facet_wrap(.~pairs, nrow = 2) +
  theme_bw() +
  theme(legend.position = "bottom",
        axis.text.x = element_text(angle = 90, hjust = 1)) +
  labs(y = "Mean instrument |exposure beta|", x = "Source study", 
       size = "n (instruments)")

gridExtra::grid.arrange(grobs = list(p_MAF, p_beta), nrow = 2)
```

## Mendelian randomisation

```{r}
#| label: MR
#| eval: FALSE

# Perform MR on each exposure-outcome pair, across each of 12 instrument sets
results_MR <- list()

for(i in 1:length(harmonised_studies)){
  
  res_list <- list()
  
  for(j in 1:length(harmonised_studies[[i]])){
    dat <- harmonised_studies[[i]][j]
    
    if(!is.na(dat[[1]]$SNP[1])){
      res <- TwoSampleMR::mr(dat[[1]])
      res$pair <- names(harmonised_studies)[i]
      res$study <- names(harmonised_studies[[i]])[j]
    } else {
      res <- data.frame(
        id.exposure = NA,
        id.outcome = NA,
        outcome = NA,
        exposure = NA,
        method = NA,
        nsnp = NA,
        b = NA,
        se = NA,
        pval = NA,
        pair = NA,
        study = NA
      )
    }
    
    res_list <- c(res_list, list(res))
  }
  res_join <- do.call("rbind", res_list)
  results_MR <- c(results_MR, list(res_join))
}

names(results_MR) <- names(harmonised_studies)

save(results_MR, file = file.path(res_dir, "results_complextrait_MR.rda"))
```

Plot of IVW estimates for each exposure-outcome pair across instrument sets

```{r}
#| fig-width: 9
#| fig-height: 12

load(file.path(res_dir, "results_complextrait_MR.rda"))

results_IVW <- do.call("rbind", results_MR) |> filter(method == "Inverse variance weighted") |>
  mutate(study = factor(study, levels = c("opengwas_common","genebass_common",
                                         "genebass_lowfreq",
                                         "genebass_rare","genebass_rare_filt",
                                         "genebass_ultrarare","genebass_ultrarare_filt",
                                         "genebass_synonymous","genebass_missense|LC","genebass_pLoF|missense|LC","genebass_pLoF",
                                         "deeprvat_genescore")))

results_IVW$instrument_type = instrument_type[as.character(results_IVW$study)]
results_IVW$instrument_class = instrument_class[as.character(results_IVW$study)]

ggplot(results_IVW, aes(x = b, y = study, colour = instrument_type)) +
  facet_wrap(.~pair, nrow = 4, scale = "free_x") +
  geom_vline(xintercept = 0, colour = "grey20", lty = 2) +
  geom_point(aes(pch = pval<0.05, size = nsnp)) +
  geom_errorbar(aes(xmin = b-1.96*se, xmax = b+1.96*se), width = 0.1) +
  scale_colour_manual(values = c("tomato","orange","darkblue")) +
  scale_shape_manual(values = c(1,19)) +
  theme_bw() +
  labs(size = "n (instruments)", colour = "Type", x = "IVW")

```

```{r}

TwoSampleMR::mr_scatter_plot(results_MR$LDL_CHD |> filter(study == "genebass_ultrarare"),
                             dat = harmonised_studies$LDL_CHD$genebass_ultrarare)


```